src/core/string.h

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2014-2015 Dropbox, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef PYSTON_CORE_STRING_H
+#define PYSTON_CORE_STRING_H
+
+#include <stdexcept>
+
+namespace pyston {
+
+class String {
+public:
+    explicit String(const std::string& s, char* _ptr) : ptr(_ptr) {
```

Comment:
I still think it's a weird mix of concerns that the storage pointer must be provided by the caller, but the constructor takes care of the copying.  It kind of feels like we just took some of the BoxedString functionality and put it in another class... I guess I'm glad we tried to make String reusable, but I just have a hard time seeing where this would be used other than BoxedString.

What if we have the BoxedString constructor do this copying, and give it a llvm_str function that returns a llvm::StringRef?